### PR TITLE
Serpent material temps

### PR DESCRIPTION
**Pull request by yardasol**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Required for Merging
- [ ] I have read the **CONTRIBUTING** document.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly. 
- [ ] My change is a source code change
  - [ ] I have added/modified tests to cover my changes
  - [ ] I have explained why my change does not require new/modified tests
- [ ] My change is a user-facing change
  - [ ] I have recorded my changes in the changelog for the upcoming release
- [ ] My change is exclusively related to the repository (e.g. GitHub actions workflows, PR/issue templates, etc.)
  - [ ] I have verified or justified that my change will work as intended.
- [ ] All new and existing tests passed.
  - [ ] CI tests pass
  - [ ] Local tests pass (including Serpent2 integration tests)
## Associated Issues and PRs

- Issue: #

## Associated Developers

- Dev: @

## Checklist for Reviewers

Reviewers should use [this link](https://arfc.github.io/manual/guides/pull_requests) to get to the 
Review Checklist before they begin their review.

### CODE REVIEW

File: saltproc/serpent_depcode.py [hunk 1]
```diff
@@ -520,13 +542,17 @@ def update_depletable_materials(self, mats, dep_end_time):
...
             for name, mat in mats.items():
-                f.write('mat  %s  %5.9E burn 1 fix %3s %4i vol %7.5E\n' %
-                        (name,
-                         -mat.density,
-                         '09c',
-                         mat.temp,
-                         mat.vol))
+                mat_card, card_volume_idx = self.burnable_material_card_data[name]
+                mat_card[2] = str(-mat.density)
+                mat_card[card_volume_idx] = "%7.5E" % mat.vol
```
[the remark this answers is not in the document]
Comment (samgdotson) [reply saltproc/serpent_depcode.py:560]: No this is string formatting

File: saltproc/serpent_depcode.py [hunk 2]
```diff
@@ -108,6 +110,20 @@ def create_runtime_matfile(self, file_lines):
...
+
+        # Get material cards
+        flines = self.read_plaintext_file(absolute_path)
+        self.get_burnable_material_card_data(flines)
+
+         # Create file with path for SaltProc rewritable iterative material file
+        shutil.copy2(absolute_path, self.runtime_matfile)
+        return [line.replace(burnable_materials_path, self.runtime_matfile) for line in file_lines]
+
+    def get_burnable_materials_file(self, file_lines):
```
[the remark this answers is not in the document]
Comment (samgdotson) [reply saltproc/serpent_depcode.py:126]: If this is a _private_ function, I will accept not having a docstring. Since users should never interact with it. In this case, the code is documentation. However, if it _is a private function_ it should have a single underscore at the beginning of the method name.

File: saltproc/serpent_depcode.py [hunk 3]
```diff
@@ -124,12 +140,18 @@ def create_runtime_matfile(self, file_lines):
...
                                   f'{self.template_input_file_path} includes '
                                   'no file with materials description')
-        # Create data directory
-        Path.mkdir(Path(self.runtime_matfile).parents[0], exist_ok=True)
+        return burnable_materials_path, absolute_path.resolve()
 
-        # Create file with path for SaltProc rewritable iterative material file
-        shutil.copy2(absolute_path, self.runtime_matfile)
-        return [line.replace(burnable_materials_path, self.runtime_matfile) for line in file_lines]
+    def get_burnable_material_card_data(self, file_lines):
```
[the remark this answers is not in the document]
Comment (samgdotson) [reply saltproc/serpent_depcode.py:145]: Same comment as above.

**Review by samgdotson — CHANGES_REQUESTED** [review]
This seems fine, just answer @abachma2's questions.

**Comment by yardasol** [comment]
@abachma2 
>For your integration tests, it looks like you've only added one test, but your function changes include if loops. Do you think you need tests to cover what happens in the if loops?

You are correct, I absolutely should.

File: saltproc/serpent_depcode.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (abachma2) [reply saltproc/serpent_depcode.py:560]: That's what I was thinking. Just wanted to make sure it wasn't hard coding any values.

File: tests/integration_tests/file_interface_serpent/test.py [hunk 4]
```diff
@@ -36,6 +36,54 @@ def test_runtime_input_from_template(serpent_depcode, msr):
     file_data = serpent_depcode.create_runtime_matfile(file_data)
     assert file_data[0].split()[-1] == '\"' + \
         serpent_depcode.runtime_matfile + '\"'
+
+    # get_burnable_material_card_data
+    _burnable_material_card_data = {'fuel':
```
Comment (abachma2) [inline tests/integration_tests/file_interface_serpent/test.py:41]: Maybe this is just a knowledge gap on my part, but why did you add the underscore here? This looks like a variable to me, not a call of that private function?

**Review by abachma2 — COMMENTED** [review]
Thanks for making these latest changes @yardasol. I have one other comment about the tests. Once you write more tests to cover the if loops in your functions I will approve and merge this PR.

File: tests/integration_tests/file_interface_serpent/test.py (hunk 4 above)
Comment (abachma2) [reply tests/integration_tests/file_interface_serpent/test.py:41]: I get that for in line 79. But even when you define it here and use it in line 80? It's still a private attribute in those locations?

**Review by abachma2 — COMMENTED** [review]
Thanks for adding the unit tests @yardasol. I think they make the test suite more robust. It looks like your unit tests are just testing that the proper error is raised. Do you also have a test in there for when the inputs have the correct information and the function should execute as expected?

File: tests/integration_tests/file_interface_serpent/test.py (hunk 4 above)
Comment (yardasol) [reply tests/integration_tests/file_interface_serpent/test.py:41]: no, this is just for naming consistency. Should I name it something else?

**Comment by yardasol** [comment]
> Do you also have a test in there for when the inputs have the correct information and the function should execute as expected?

Yes, that machinery is added to the integration test `test_file_interface_serpent`

File: tests/integration_tests/file_interface_serpent/test.py (hunk 4 above)
Comment (abachma2) [reply tests/integration_tests/file_interface_serpent/test.py:41]: I would recommend changing the same some when you're not calling the attribute. It was a little confusing to me to have the two things you're comparing be named the same thing. I like to use the `exp` and `obs` (as in expected and observed) that Katy uses in her book for my tests. But your choice as to what you change it to.

**Comment by abachma2** [comment]
> > Do you also have a test in there for when the inputs have the correct information and the function should execute as expected?
> 
> Yes, that machinery is added to the integration test `test_file_interface_serpent`

And you think that having that only in an integration test is sufficient? It doesn't need its own unit test? I'm not expecting a certain answer, I'm genuinely asking.

**Comment by yardasol** [comment]
> And you think that having that only in an integration test is sufficient? It doesn't need its own unit test?

It should be sufficient. As long as the user provides correct input paths none of the errors should be raised, and the code will run correctly. Maybe @munkm can chime in?

File: tests/integration_tests/file_interface_serpent/test.py (hunk 4 above)
Comment (yardasol) [reply tests/integration_tests/file_interface_serpent/test.py:41]: Okay I'll do this

**Comment by yardasol** [comment]
@samgdotson bump

**Comment by munkm** [comment]
ok @yardasol and @abachma2 I took a look. I think generally the integration test is probably good enough for this case. However, having a unit test that checks the specific function like Amanda brought up would be even better. Users are good at generating all sorts of weird corner cases and this integration test isn't going to capture that (or any improvements to the tooling that detects the input paths). Are errors raised consistently for all sorts of weird iterations of paths (I don't see that tested here)? Are the expected errors raised? If a user comes up with a case where they enter a path wrong and the error isn't raised, where would they modify this integration test to include it? I don't see an obvious spot in `test_file_interface_serpent`. I don't see an assertion that an error is raised with a wrong configuration either. Am I missing it?

**Comment by yardasol** [comment]
@munkm 
> Are errors raised consistently for all sorts of weird iterations of paths (I don't see that tested here)?

If a path is bad, Python should raise a FileNotFoundError. My personal opinion is that it is out-of-scope to try to test fundamental Python machinery in a program built on top of it.

> Are the expected errors raised?

See [this test](https://github.com/arfc/saltproc/pull/178/files#diff-f0fddb7361e93210e82ae15153a3b4d39888d25e1f40c542bc97fd594cb64cdfR40).

> If a user comes up with a case where they enter a path wrong and the error isn't raised, where would they modify this integration test to include it?

Again, this is more of an issue with python itself than it is with SaltProc.

> I don't see an assertion that an error is raised with a wrong configuration either. Am I missing it?

I'm not sure what you mean by this. Are you talking about the return message? Or are you talking about assertions than a particular error is raised? If the latter, see the above link.

**Comment by yardasol** [comment]
@munkm bump

**Review by samgdotson — APPROVED** [review]
This seems fine, apologies for the delay.

**Comment by munkm** [comment]
I'm good with this merge now that there's an open issue! Thanks for doing that @samgdotson

**Comment by munkm** [comment]
But please next time check with the other people who have started review before merging. It wasn't clear if @abachma2 had approved yet.
